core/src/main/java/org/jruby/runtime/invokedynamic/MathLinker.java

```diff
@@ -141,84 +152,73 @@ public static IRubyObject fixnumOperator(ThreadContext context, IRubyObject call
 
-        if (LOG_BINDING) LOG.debug(name + "\tFixnum boolean operation at site #" + site.siteID() + " (" + site.file() + ":" + site.line() + ") bound directly");
+        RubyClass classFixnum = context.runtime.getFixnum();
+        int gen = classFixnum.getGeneration();
+        CacheEntry entry = searchWithCache(operator, caller, self.getMetaClass(), site);
+        if (entry == null || !entry.method.isBuiltin()) gen--; // invalid - fallback on slow-path
+
+        MethodHandle test = FIXNUM_TEST_WITH_GEN; // NOTE: since we invalidate on Integer change FIXNUM_TEST should be enough?!
+        test = insertArguments(test, 1, classFixnum, gen);
+        test = permuteArguments(test, methodType(boolean.class, ThreadContext.class, IRubyObject.class, IRubyObject.class), ARG_2_TO_0);
+
+        if (LOG_BINDING) LOG.debug(name + "\tFixnum boolean operation at site #" + site.siteID + " (" + site.file() + ":" + site.line() + ") bound directly");
         
         // confirm it's a Fixnum
         target = guardWithTest(test, target, fallback);
-        
-        // check Fixnum reopening
-        target = ((SwitchPoint)context.runtime.getFixnumInvalidator().getData())
-                .guardWithTest(target, fallback);
-        
+        target = ((SwitchPoint) classFixnum.getInvalidator().getData()).guardWithTest(target, fallback);
         site.setTarget(target);
         
-        return (Boolean)target.invokeWithArguments(context, caller, self);
+        return (Boolean) target.invokeWithArguments(context, caller, self);
     }
-    
-    public static IRubyObject fixnumOperatorFail(ThreadContext context, IRubyObject caller, IRubyObject self, JRubyCallSite site, RubyFixnum value) throws Throwable {
-        String operator = site.name();
-        RubyClass selfClass = InvokeDynamicSupport.pollAndGetClass(context, self);
-        CacheEntry entry = site.entry;
-        
-        if (entry.typeOk(selfClass)) {
-            return entry.method.call(context, self, entry.sourceModule, operator, value);
-        } else {
-            entry = selfClass.searchWithCache(operator);
-            if (InvokeDynamicSupport.methodMissing(entry, site.callType(), operator, caller)) {
-                return InvokeDynamicSupport.callMethodMissing(entry, site.callType(), context, self, operator, value);
-            }
-            site.entry = entry;
-            return entry.method.call(context, self, entry.sourceModule, operator, value);
-        }
+
+    static boolean fixnumTest(IRubyObject self) {
+        return self instanceof RubyFixnum;
     }
 
-    public static boolean fixnumBooleanFail(ThreadContext context, IRubyObject caller, IRubyObject self, JRubyCallSite site, RubyFixnum value) throws Throwable {
+    static boolean fixnumTestWithGen(IRubyObject self, RubyClass klass, final int gen) {
+        return self instanceof RubyFixnum && klass.getGeneration() == gen;
```

Comment:
This is a nice addition but ideally the indy way would be to use a switch point rather than a generation check. The generation check will always require traversing the class and the field, where the switch point is compiled as a safe point with basically zero overhead.

This is fine for now; we have continued to flip the generation since many other places can't use switch point, but the latter will optimize better.

Reply:
this was an addition on top of the method invalidation, as explained: https://github.com/jruby/jruby/pull/4736/files#r297135999
we can get rid of it and see how it goes.